# **Last Name:** SIDIBE

# **First Name:** Moussa

In [381]:
from jyquickhelper import add_notebook_menu

add_notebook_menu()

##  Packages requires

In [382]:
import pandas as pd
import json
from collections import Counter,OrderedDict
import time
import os

## Data loading

In [383]:
artist_choices=['Zaz','Oumou Sangaré','Kendrick Lamar',
                  'Pink','The Killers','Muse','Eminem',
                  'Taylor Swift','Harry Styles','Lara Fabian',
                  'Laurent Gamier'
                 ]

In [384]:
def open_file_json_and_framed(json_doc):
    
    with open(json_doc,'r') as f:
        json_file=json.load(f)

    df=pd.DataFrame(json_file['playlists'],columns=json_file['playlists'][0].keys())
    
    return df

df_playlists=open_file_json_and_framed("mpd.slice.0-999.json")

In [385]:
df_playlists.head()

name collaborative  pid  modified_at  num_tracks  num_albums  \
0        Throwbacks         false    0   1493424000          52          47   
1  Awesome Playlist         false    1   1506556800          39          23   
2           korean          false    2   1505692800          64          51   
3               mat         false    3   1501027200         126         107   
4               90s         false    4   1401667200          17          16   

   num_followers                                             tracks  \
0              1  [{'pos': 0, 'artist_name': 'Missy Elliott', 't...   
1              1  [{'pos': 0, 'artist_name': 'Survivor', 'track_...   
2              1  [{'pos': 0, 'artist_name': 'Hoody', 'track_uri...   
3              1  [{'pos': 0, 'artist_name': 'Camille Saint-Saën...   
4              2  [{'pos': 0, 'artist_name': 'The Smashing Pumpk...   

   num_edits  duration_ms  num_artists  
0          6     11532414           37  
1          5     11656470           21  
2         18     14039958           31  
3          4     28926058           86  
4          7      4335282           16

## Counting the number of playlist by artist

In [386]:
def initialize_dict_number_playlist_by_artist(artist_choices):
    
        '''

         This function is made to initialise dictionnary by zero as values

        Input:
            artist_choices: list of artist choices to display some informations

        Return:

            nnumber_playslist_by_artist: a dictionnary that contain as keys artist_choices and as values 0 as default value

        '''

        number_playslist_by_artist=dict()

        for artist in artist_choices:

            number_playslist_by_artist[artist]=0

        return number_playslist_by_artist



def count_number_playlist_by_artist(df_playlists,artist_choices):
    
        '''

        This function take two argument and return number of track by an artist in artist_choices

        Input:

            df_playlists: DataFrame of playlists, which playlists contain many tracks

            artist_choices: list of artist choices to display some informations

        Output:

            number_track_by_artist: a dictionnary of number of playlist by artist

        '''

        number_playslist_by_artist=initialize_dict_number_playlist_by_artist(artist_choices)

        for j,track in enumerate(df_playlists['tracks']):

            artist_playlist=[]

            for i in range(df_playlists['num_tracks'][j]):

                artist_playlist.append(track[i]['artist_name'])

            for artist in artist_choices:

                if artist in set(artist_playlist):

                    number_playslist_by_artist[artist]+=1

        return number_playslist_by_artist

In [387]:
start_time=time.time()

number_playslist_by_artist=count_number_playlist_by_artist(df_playlists,artist_choices)

end_time=time.time()

print(number_playslist_by_artist)

print('\n Operation done in %f seconds'%(end_time-start_time))

{'Zaz': 0, 'Oumou Sangaré': 0, 'Kendrick Lamar': 121, 'Pink': 0, 'The Killers': 40, 'Muse': 14, 'Eminem': 91, 'Taylor Swift': 36, 'Harry Styles': 30, 'Lara Fabian': 0, 'Laurent Gamier': 0}

 Operation done in 0.064609 seconds


## Counting the number of track by artist

In [388]:
def initialize_dict_number_track_by_artist(artist_choices):

        '''
        This function is made to initialise dictionnary by zero as values

        Input:
            artist_choices: list of artist choices to display some informations

        Return:

            number_track_by_artist: a dictionnary that contain as key artist_choices and as values 0 as default value

        '''

        number_track_by_artist=dict()

        for artist in artist_choices:

            number_track_by_artist[artist]=0

        return number_track_by_artist


def count_number_track_by_artist(df_playlists,artist_choices):
    
    '''
    
    This function take two argument and return number of track by an artist in artist_choices
    
    Input:
        
        df_playlists: DataFrame of playlists, which playlists contain many tracks
        
        artist_choices: list of artist choices to display some informations
        
    Output:
        
        number_track_by_artist: a dictionnary of number of tracks by artist
        
    '''
    
    
    number_track_by_artist=initialize_dict_number_track_by_artist(artist_choices)
    
    for j,track in enumerate(df_playlists['tracks']):

        artist_track=[]

        for i in range(df_playlists['num_tracks'][j]):

            artist_track.append(track[i]['artist_name'])

        track_by_artist=Counter(artist_track)
        
        for artist in artist_choices:

            if artist in track_by_artist.keys():
                
                number_track_by_artist[artist]+=track_by_artist[artist]
                
    return number_track_by_artist

In [389]:
start_time=time.time()

number_track_by_artist=count_number_track_by_artist(df_playlists,artist_choices)

end_time=time.time()

print(number_track_by_artist)

print("\n Operation done in %f second(s)"%(end_time-start_time))

{'Zaz': 0, 'Oumou Sangaré': 0, 'Kendrick Lamar': 385, 'Pink': 0, 'The Killers': 55, 'Muse': 33, 'Eminem': 332, 'Taylor Swift': 64, 'Harry Styles': 74, 'Lara Fabian': 0, 'Laurent Gamier': 0}

 Operation done in 0.048727 second(s)


## Sum of duration of songs in playlists by artist

In [390]:
def initialize_dict_duration_amount_by_artist(artist_choices):
    
        '''
        This function is made to initialise dictionnary by zero as values

        Input:
            artist_choices: list of artist choices to display some informations

        Return:
            return a dictionnary that contain as key artist_choices and as values 0

        '''

        amount_duration_by_artist=dict()

        for artist in artist_choices:

            amount_duration_by_artist[artist]=0

        return amount_duration_by_artist


def sum_amount_duration_by_artist(df_playlists,artist_choices):
    
        '''
        This function take two argument and return sum of duration of apparition of an artist in artist_choices

        Input:

            df_playlists: DataFrame of playlists, which playlists contain many tracks

            artist_choices: list of artist choices to display some informations
        Output:

            amount_duration_by_artist: a dictionnary of amount duration by artist in a pretty format

        '''

        amount_duration_by_artist=initialize_dict_duration_amount_by_artist(artist_choices)

        for j,track in enumerate(df_playlists['tracks']):

            duration_artist_track={}

            for i in range(df_playlists['num_tracks'][j]):

                name=track[i]['artist_name']

                if name in artist_choices:

                    amount_duration_by_artist[name]+=track[i]['duration_ms']
        
        return amount_duration_by_artist
    
    
def pretty_print_amount(amount_duration_by_artist):
    
    beauty_amount_duration_by_artist=dict()
    
    for artist in artist_choices:
        amount=amount_duration_by_artist[artist]
        beauty_amount_duration_by_artist[artist]=str(int(amount*10**(-3)//60))+' min '+str(int(amount*10**(-3)//3600))

    print(beauty_amount_duration_by_artist)

In [391]:
start_time=time.time()

amount_duration_by_artist=sum_amount_duration_by_artist(df_playlists,artist_choices)

end_time=time.time()

pretty_print_amount(amount_duration_by_artist)

print("\n Operation done in %f second(s)"%(end_time-start_time))

{'Zaz': '0 min 0', 'Oumou Sangaré': '0 min 0', 'Kendrick Lamar': '1629 min 27', 'Pink': '0 min 0', 'The Killers': '208 min 3', 'Muse': '157 min 2', 'Eminem': '1623 min 27', 'Taylor Swift': '248 min 4', 'Harry Styles': '299 min 4', 'Lara Fabian': '0 min 0', 'Laurent Gamier': '0 min 0'}

 Operation done in 0.043510 second(s)


## Generalisation of the built on 1000 playlists on a Million playlists

In [392]:
def sum_dict_by_key(dict1,dict2):
    
    """
    
    This function is use to sum dictionnaries by the key extend with new keys
    
    Input:
    dict1,dict2: dictionnaries
    
    Return:
    
    Dictionnary of two dictionnary, with updating values of keys that exist in both dicts and added 
    
    others as they was
    
    """
    
    dict3=dict2
    
    for key in dict1.keys():
        
        if key in dict2.keys():
            
            dict3[key]+=dict1[key]
        
        else:
            
            dict3[key]=dict1[key]
    
    return dict3

In [393]:
def divide_and_apply(function,artist_choices):
    
    """
    This function take as parameters another functions built on small dataset and apply to larger
    
    Input
    
    function: a function that will be apply
    
    artist_choices: selected artists 
    
    Return:
    
    A dictionnary that contains the output aggregated of function on 1000 datasets of 1000 playlists
    
    """
    
    
    files = ['./Million/data/'+f for f in os.listdir('./Million/data') if f.endswith('.json')]
    
    
    final_result={}
    
    for file_doc in files:
        
        df=open_file_json_and_framed(file_doc)
        
        
        temp_result=function(df,artist_choices)
        
        final_result=sum_dict_by_key(final_result,temp_result)
        
    return final_result

In [394]:
start_time=time.time()

number_playlist_by_artist=divide_and_apply(count_number_playlist_by_artist,artist_choices)

number_track_by_artist=divide_and_apply(count_number_track_by_artist,artist_choices)

amount_duration_by_artist=divide_and_apply(sum_amount_duration_by_artist,artist_choices)

end_time=time.time()

In [395]:
print("Number of playlists by artists are:\n", number_playslist_by_artist)

print("="*114)

print("Number of track by artists are:\n", number_track_by_artist)

print("="*114)

print("Sum of duration by artists are:\n")

pretty_print_amount(amount_duration_by_artist)


print("="*114)

duration=end_time-start_time

print("All three operations was done in %d min:%f secondes"%(duration//60,duration%60))

Number of playlists by artists are:
 {'Zaz': 0, 'Oumou Sangaré': 0, 'Kendrick Lamar': 121, 'Pink': 0, 'The Killers': 40, 'Muse': 14, 'Eminem': 91, 'Taylor Swift': 36, 'Harry Styles': 30, 'Lara Fabian': 0, 'Laurent Gamier': 0}
Number of track by artists are:
 {'Zaz': 1041, 'Oumou Sangaré': 153, 'Kendrick Lamar': 353624, 'Pink': 0, 'The Killers': 81685, 'Muse': 35807, 'Eminem': 294667, 'Taylor Swift': 105366, 'Harry Styles': 32003, 'Lara Fabian': 314, 'Laurent Gamier': 0}
Sum of duration by artists are:

{'Zaz': '3507 min 58', 'Oumou Sangaré': '813 min 13', 'Kendrick Lamar': '1480127 min 24668', 'Pink': '0 min 0', 'The Killers': '322017 min 5366', 'Muse': '161592 min 2693', 'Eminem': '1398647 min 23310', 'Taylor Swift': '409438 min 6823', 'Harry Styles': '143459 min 2390', 'Lara Fabian': '1400 min 23', 'Laurent Gamier': '0 min 0'}
All three operations was done in 16 min:50.986956 secondes


## Suggest a mixable way of two playlist

- The idea, that I developped for this question is to develop an algorithm to find common artist between two playlist by using intersection between both set of artist of playlist.
  
  This solution will be useful to estimate similarities between two playlists and the singers in common can be used to move from a playlist to another. This is the potential solution of the question
  
- With the purpose to generalize to all dataset, I find useful to compute the number of common artist between all playlists and ordered them to find the best ways to move between playlists. With this solution we can construct an organigram of playlists with best similar to low similarity and we dropout ones we haven't none artist in common then movement is not possible.

In [396]:
def get_list_tracks(pid1,pid2):
    
    """
    
    This function return intersection between two set of artists of different playlists
    
    accessed with pid.
    
    Input:
    
    pid1: playlist and track id for the first playlist
    
    pid2: playlist and track id for the second playlist
    
    Return:
    
    set of artist names that are common to both playlists
    
    """
    
    a=[[],[]]

    for j,pid in enumerate([pid1,pid2]):

        for i in range(df_playlists['num_tracks'][pid]):

            a[j].append(df_playlists['tracks'][pid][i]['artist_name'])
        
            
    return set(a[0]).intersection(set(a[1]))

In [397]:
def recommadantion(df):
    
    """
    
    Input:
    
    df: DataFrame of playlists
    
    
    Return:
    
    A list of possible movement between playlists oredered from most similarity to lower ...
    
    """
    
    pids_with_nb_common_artist={}
    
    for pid1 in df.pid:
        
        for pid2 in df.pid:
            
            future_key1=str(pid1)+"---> "+str(pid2)
            
            future_key2=str(pid2)+"---> "+str(pid1)
            
            keys=pids_with_nb_common_artist.keys()
            
            if future_key1 not in keys and future_key2 not in keys and pid1!=pid2:
            
                pids_with_nb_common_artist[str(pid1)+"---> "+str(pid2)]=len(get_list_tracks(pid1,pid2))
                
    Liste_of_graphe_sorted=sorted(pids_with_nb_common_artist.items(), key=lambda t: t[1])

    Final_liste={}

    for key, value in Liste_of_graphe_sorted:

            if value!=0:
            
                Final_liste[key]=value
                
    return OrderedDict(reversed(list(Final_liste.items())))

In [400]:
start_time=time.time()

Liste_of_graphe=recommadantion(df_playlists)

end_time=time.time()

duration=end_time-start_time

print("Recommandation operate in %d min:%f secondes"%(duration//60,duration%60))

Recommandation operate in 7 min:14.196659 secondes


In [401]:
list(Liste_of_graphe)[:10]

['559---> 980',
 '123---> 782',
 '559---> 907',
 '559---> 843',
 '123---> 815',
 '95---> 559',
 '412---> 879',
 '412---> 803',
 '782---> 815',
 '559---> 835']

Of course, in this question we have use 1000 playlists to avoid Million.

We figure out that the playlists 559 <---> 980 and 123 <---> 782 are the most similar suits by others. Then a music fan who play playlist 559 will more probably liked 980 ...

# THANK YOU !!!